### PR TITLE
[DAG] fold avgu(zext(x), zext(y)) -> zext(avgu(x, y))

### LINKED ISSUE #86301: [DAG] Fold AVGU(ZEXT(X),ZEXT(Y)) -> ZEXT(AVGU(X,Y))
This works for AVGCEILU and AVGFLOORU: https://alive2.llvm.org/ce/z/ea-m5Y

If the pre-extended AVGCEILU/AVGFLOORU node is legal, then we should perform the operation with the shorter width and then extend afterward.

### PR DESCRIPTION
**Pull request by c8ef**
close: #86301

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 1]
```diff
@@ -5236,6 +5236,21 @@ SDValue DAGCombiner::visitAVG(SDNode *N) {
     return DAG.getNode(ISD::SRL, DL, VT, X,
                        DAG.getShiftAmountConstant(1, VT, DL));
 
+  // fold avgu(zext(x), zext(y)) -> zext(avgu(x, y))
+  SDValue A;
+  SDValue B;
+  if (hasOperation(ISD::AVGFLOORU, VT) &&
+      sd_match(N, m_c_BinOp(ISD::AVGFLOORU, m_ZExt(m_Value(A)),
+                            m_ZExt(m_Value(B))))) {
+    SDValue AvgFloorU = DAG.getNode(ISD::AVGFLOORU, DL, A.getValueType(), A, B);
```
Comment (jayfoad) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:5245]: Don't you also need to check that A and B are both being extended _from_ the same size?

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 2]
```diff
@@ -5236,6 +5236,21 @@ SDValue DAGCombiner::visitAVG(SDNode *N) {
     return DAG.getNode(ISD::SRL, DL, VT, X,
                        DAG.getShiftAmountConstant(1, VT, DL));
 
+  // fold avgu(zext(x), zext(y)) -> zext(avgu(x, y))
+  SDValue A;
+  SDValue B;
+  if (hasOperation(ISD::AVGFLOORU, VT) &&
```
Comment (jayfoad) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:5242]: Should be testing hasOperation with the new type, A.getValueType() ?

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 3]
```diff
@@ -5236,6 +5236,23 @@ SDValue DAGCombiner::visitAVG(SDNode *N) {
     return DAG.getNode(ISD::SRL, DL, VT, X,
                        DAG.getShiftAmountConstant(1, VT, DL));
 
+  // fold avgu(zext(x), zext(y)) -> zext(avgu(x, y))
+  SDValue A;
+  SDValue B;
+  if (sd_match(N, m_c_BinOp(ISD::AVGFLOORU, m_ZExt(m_Value(A)),
```
Comment (dtcxzyw) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:5242]:
```suggestion
  if (sd_match(N, m_BinOp(ISD::AVGFLOORU, m_ZExt(m_Value(A)),
```

Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:5242]: AVGFLOORU/AVGCEILU are commutative so m_c_BinOp should work no?

Comment (dtcxzyw) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:5242]: We don't need commutative matchers as both LHS and RHS are the same.

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 4]
```diff
@@ -5236,6 +5236,23 @@ SDValue DAGCombiner::visitAVG(SDNode *N) {
     return DAG.getNode(ISD::SRL, DL, VT, X,
                        DAG.getShiftAmountConstant(1, VT, DL));
 
+  // fold avgu(zext(x), zext(y)) -> zext(avgu(x, y))
+  SDValue A;
+  SDValue B;
+  if (sd_match(
+          N, m_BinOp(ISD::AVGFLOORU, m_ZExt(m_Value(A)), m_ZExt(m_Value(B)))) &&
+      A.getValueType() == B.getValueType() &&
```
Comment (dtcxzyw) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:5244]: Can you add a negative test for `A.getValueType() != B.getValueType()` case?

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!

**Review by RKSimon — APPROVED** [review]
LGTM - cheers

**Comment by jayfoad** [comment]
Any plans to do avgs(sext(x), sext(y)) -> sext(avgs(x, y)) ?

**Comment by RKSimon** [comment]
> Any plans to do avgs(sext(x), sext(y)) -> sext(avgs(x, y)) ?

IIRC the constraints on that pattern aren't as easy as the unsigned cases.

**Comment by jayfoad** [comment]
> > Any plans to do avgs(sext(x), sext(y)) -> sext(avgs(x, y)) ?
> 
> IIRC the constraints on that pattern aren't as easy as the unsigned cases.

Really? My instinct is that it should Just Work, but I have not thought about it deeply.

**Comment by RKSimon** [comment]
> > > Any plans to do avgs(sext(x), sext(y)) -> sext(avgs(x, y)) ?
> > 
> > 
> > IIRC the constraints on that pattern aren't as easy as the unsigned cases.
> 
> Really? My instinct is that it should Just Work, but I have not thought about it deeply.

I'm miss-remembering - it looks like it just needs an explicit freeze in the alive test where the unsigned cases didn't https://alive2.llvm.org/ce/z/qgp7bF

**Comment by dtcxzyw** [comment]
> > > > Any plans to do avgs(sext(x), sext(y)) -> sext(avgs(x, y)) ?
> > > 
> > > 
> > > IIRC the constraints on that pattern aren't as easy as the unsigned cases.
> > 
> > 
> > Really? My instinct is that it should Just Work, but I have not thought about it deeply.
> 
> I'm miss-remembering - it looks like it just needs an explicit freeze in the alive test where the unsigned cases didn't https://alive2.llvm.org/ce/z/qgp7bF

I don't think we need freeze here because this fold doesn't introduce multiple uses as `xor + or` does. BTW, we already gave up on correct undef semantics in https://github.com/llvm/llvm-project/pull/90097.

**Pull request merged**